massa-versioning-worker/src/versioning_factory.rs

```diff
@@ -0,0 +1,402 @@
+use massa_time::MassaTime;
+use std::collections::BTreeMap;
+use std::iter;
+use thiserror::Error;
+
+use crate::versioning::{MipComponent, MipState, MipStateTypeId, MipStore};
+
+/// Factory error
+#[allow(missing_docs)]
+#[derive(Error, Debug)]
+pub enum FactoryError {
+    #[error("Unknown version, cannot build obj with version: {0}")]
+    UnknownVersion(u32),
+    #[error("Version: {0} is in versioning store but not yet implemented")]
+    UnimplementedVersion(u32), // programmer error, forgot to implement new version in factory traits
+    #[error("Version: {0} is not active yet")]
+    OnStateNotReady(u32),
+    #[error("Could not create object of type {0}: {1}")]
+    OnCreate(String, String),
+}
+
+#[derive(Clone)]
+/// Strategy to use when creating a new object from a factory
+pub enum FactoryStrategy {
```

Comment:
Maybe rename FactoryStrategy::Best to LatestActive and FactoryStrategy::At() to ActiveAt() to be more explicit?

Reply:
This adds some 'State info' for the dev that will implement a factory and I'm not a big fan of this... Maybe we can add some doc in Factory trait for create method to make it clear the dev?

Reply:
done